// @armoriq/sdk/dist/exceptions.js
"use strict";
/**
 * Custom exceptions for ArmorIQ SDK.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationException = exports.PolicyHoldException = exports.PolicyBlockedException = exports.DelegationException = exports.MCPInvocationException = exports.TokenExpiredException = exports.IntentMismatchException = exports.InvalidTokenException = exports.ArmorIQException = void 0;
/**
 * Base exception for all ArmorIQ SDK errors.
 */
class ArmorIQException extends Error {
    constructor(message) {
        super(message);
        this.name = 'ArmorIQException';
        Object.setPrototypeOf(this, ArmorIQException.prototype);
    }
}
exports.ArmorIQException = ArmorIQException;
/**
 * Raised when an intent token is invalid.
 *
 * This can occur due to:
 * - Invalid signature
 * - Token expiration
 * - Token revocation
 * - Malformed token structure
 */
class InvalidTokenException extends ArmorIQException {
    tokenId;
    constructor(message, tokenId) {
        super(message);
        this.name = 'InvalidTokenException';
        this.tokenId = tokenId;
        Object.setPrototypeOf(this, InvalidTokenException.prototype);
    }
}
exports.InvalidTokenException = InvalidTokenException;
/**
 * Raised when an action doesn't match the original intent plan.
 *
 * This occurs when trying to execute an action that was not
 * included in the canonicalized plan used to generate the token.
 */
class IntentMismatchException extends ArmorIQException {
    action;
    planHash;
    constructor(message, action, planHash) {
        super(message);
        this.name = 'IntentMismatchException';
        this.action = action;
        this.planHash = planHash;
        Object.setPrototypeOf(this, IntentMismatchException.prototype);
    }
}
exports.IntentMismatchException = IntentMismatchException;
/**
 * Raised when an intent token has expired.
 *
 * Tokens have a validity period set during issuance.
 * This exception is raised when attempting to use an expired token.
 */
class TokenExpiredException extends InvalidTokenException {
    expiredAt;
    constructor(message, tokenId, expiredAt) {
        super(message, tokenId);
        this.name = 'TokenExpiredException';
        this.expiredAt = expiredAt;
        Object.setPrototypeOf(this, TokenExpiredException.prototype);
    }
}
exports.TokenExpiredException = TokenExpiredException;
/**
 * Raised when an MCP action invocation fails.
 *
 * This can occur due to:
 * - MCP server unavailable
 * - Action not found
 * - Invalid parameters
 * - Proxy verification failure
 */
class MCPInvocationException extends ArmorIQException {
    mcp;
    action;
    statusCode;
    constructor(message, mcp, action, statusCode) {
        super(message);
        this.name = 'MCPInvocationException';
        this.mcp = mcp;
        this.action = action;
        this.statusCode = statusCode;
        Object.setPrototypeOf(this, MCPInvocationException.prototype);
    }
}
exports.MCPInvocationException = MCPInvocationException;
/**
 * Raised when agent delegation fails.
 *
 * This can occur due to:
 * - Target agent unavailable
 * - Trust delegation rejected
 * - Invalid subtask structure
 */
class DelegationException extends ArmorIQException {
    targetAgent;
    delegationId;
    statusCode;
    constructor(message, targetAgent, delegationId, statusCode) {
        super(message);
        this.name = 'DelegationException';
        this.targetAgent = targetAgent;
        this.delegationId = delegationId;
        this.statusCode = statusCode;
        Object.setPrototypeOf(this, DelegationException.prototype);
    }
}
exports.DelegationException = DelegationException;
/**
 * Raised when a policy enforcement blocks the action.
 */
class PolicyBlockedException extends ArmorIQException {
    enforcementAction;
    reason;
    metadata;
    constructor(message, enforcementAction, reason, metadata) {
        super(message);
        this.name = 'PolicyBlockedException';
        this.enforcementAction = enforcementAction;
        this.reason = reason;
        this.metadata = metadata;
        Object.setPrototypeOf(this, PolicyBlockedException.prototype);
    }
}
exports.PolicyBlockedException = PolicyBlockedException;
/**
 * Raised when a policy enforcement holds the action for approval.
 */
class PolicyHoldException extends ArmorIQException {
    delegationContext;
    metadata;
    constructor(message, delegationContext, metadata) {
        super(message);
        this.name = 'PolicyHoldException';
        this.delegationContext = delegationContext;
        this.metadata = metadata;
        Object.setPrototypeOf(this, PolicyHoldException.prototype);
    }
}
exports.PolicyHoldException = PolicyHoldException;
/**
 * Raised when SDK configuration is invalid.
 *
 * This can occur due to:
 * - Missing required configuration
 * - Invalid endpoint URLs
 * - Missing credentials
 */
class ConfigurationException extends ArmorIQException {
    constructor(message) {
        super(message);
        this.name = 'ConfigurationException';
        Object.setPrototypeOf(this, ConfigurationException.prototype);
    }
}
exports.ConfigurationException = ConfigurationException;
