// @armoriq/sdk/dist/index.js
"use strict";
/**
 * ArmorIQ SDK - Build Secure AI Agents
 *
 * A TypeScript SDK for building AI agents with cryptographic intent verification.
 * Provides simple APIs for plan capture, token management, and secure MCP
 * tool invocation with built-in security.
 *
 * @author ArmorIQ Team <license@armoriq.io>
 * @version 0.2.6
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMAIL = exports.AUTHOR = exports.VERSION = exports.IntentToken = exports.PolicyHoldException = exports.PolicyBlockedException = exports.ConfigurationException = exports.DelegationException = exports.TokenExpiredException = exports.MCPInvocationException = exports.IntentMismatchException = exports.InvalidTokenException = exports.ArmorIQException = exports.ArmorIQClient = void 0;
var client_1 = require("./client");
Object.defineProperty(exports, "ArmorIQClient", { enumerable: true, get: function () { return client_1.ArmorIQClient; } });
var exceptions_1 = require("./exceptions");
Object.defineProperty(exports, "ArmorIQException", { enumerable: true, get: function () { return exceptions_1.ArmorIQException; } });
Object.defineProperty(exports, "InvalidTokenException", { enumerable: true, get: function () { return exceptions_1.InvalidTokenException; } });
Object.defineProperty(exports, "IntentMismatchException", { enumerable: true, get: function () { return exceptions_1.IntentMismatchException; } });
Object.defineProperty(exports, "MCPInvocationException", { enumerable: true, get: function () { return exceptions_1.MCPInvocationException; } });
Object.defineProperty(exports, "TokenExpiredException", { enumerable: true, get: function () { return exceptions_1.TokenExpiredException; } });
Object.defineProperty(exports, "DelegationException", { enumerable: true, get: function () { return exceptions_1.DelegationException; } });
Object.defineProperty(exports, "ConfigurationException", { enumerable: true, get: function () { return exceptions_1.ConfigurationException; } });
Object.defineProperty(exports, "PolicyBlockedException", { enumerable: true, get: function () { return exceptions_1.PolicyBlockedException; } });
Object.defineProperty(exports, "PolicyHoldException", { enumerable: true, get: function () { return exceptions_1.PolicyHoldException; } });
var models_1 = require("./models");
Object.defineProperty(exports, "IntentToken", { enumerable: true, get: function () { return models_1.IntentToken; } });
exports.VERSION = '0.2.6';
exports.AUTHOR = 'ArmorIQ Team';
exports.EMAIL = 'license@armoriq.io';
